src/commons/editor/useComments.tsx

```diff
@@ -0,0 +1,261 @@
+import { require as acequire, Ace } from 'ace-builds';
+import * as React from 'react';
+import ReactDOM from 'react-dom';
+
+import { groupBy, map, omit, sortBy, values, keyBy } from 'lodash';
+
+import { EditorHook } from './Editor';
+import Comments, { IComment } from './Comments';
+import { v1 as uuidv1 } from 'uuid';
+import { EventEmitter } from 'events';
+const LineWidgets = acequire('ace/line_widgets').LineWidgets;
+
+// Inferred from: https://github.com/ajaxorg/ace/blob/master/lib/ace/ext/error_marker.js#L129
+interface IWidget {
+  row: number;
+  fixedWidth: boolean;
+  coverGutter: boolean;
+  el: Element;
+  type: string;
+}
+
+interface ILineManager {
+  attach: (editor: Ace.Editor) => void;
+  addLineWidget: (widget: IWidget) => void;
+  removeLineWidget: (widget: IWidget) => void;
+}
+
+export interface CommentAPI {
+  // Assumes commentId is immutable. Do not update it unless explicitly via API.
+  updateComment: (...comments: IComment[]) => void;
+  updateCommentsBeingEdited: (...comments: IComment[]) => void;
+  removeComment: (comment: IComment) => void;
+  removeCommentEdit: (comment: IComment) => void;
+  // Probably need a replaceId for comments to update when the server does.
+
+  isUnsubmittedComment: (comment: IComment) => boolean;
+}
+
+/*  NOTE ON INTEGRATION WITH SERVER SIDE:
+
+All the data is bound in this component so that it can easily be updated from server side.
+
+I'm guessing that useCallback + ref can be used to pass real time updates in.
+
+Specifically:
+
+```
+
+const onExternalDataRef = React.useState((data: dataT) => {});
+const onExternalData = React.useCallback( (data: dataT) => {
+    // ...  setState( ... ) ... 
+}, [comments, etc.]);
+onExternalDataRef.current = onExternalData;
+
+React.useEffect(() => {
+    externalService.on('data', (data) => onExternalDataRef.current(data));
+}, [onExternalDataRef])
+
+```
+
+Also note that for the comment ID, the server should not trust the 
+ID being sent to be globally unique.
+- Not cryptographically secure
+- Even if it was, the user can just send a known collision.
+
+Discard everything except following:
+- id (for update purposes, but ignore it for new comments)
+- text
+- linenum
+
+*/
+
+const useComments: EditorHook = (
+  inProps,
+  outProps,
+  keyBindings,
+  reactAceRef,
+  contextMenuHandlers
+) => {
+  const [comments, setComments] = React.useState({} as { [id: string]: IComment });
+  // Comments being edited are copies of existing comments.
+  const [commentsBeingEdited, _setCommentsBeingEdited] = React.useState(
+    {} as { [id: string]: IComment }
```

Comment:
We might want to type `{ [id: string]: IComment }` as `ICommentMap` or something, otherwise it gets a bit unwieldy to read after a while.